### DISCUSSION TITLE
Smart Contract Implementation

### DISCUSSION THREAD

**Discussion by kanaagawa** [1Hive/Vested#19] [Ideas]
After a review of the existing solutions described in #1 my opinion is to work on an implementation reusing the [MerkleDistributor](https://github.com/Uniswap/merkle-distributor/blob/master/contracts/MerkleDistributor.sol) as the `Offer` contract and extend it to be an ERC-721 token. As discussed this allow vesting contracts to be transfer as an NFT.

During user [claims](https://github.com/Uniswap/merkle-distributor/blob/master/contracts/MerkleDistributor.sol#L43) we should:
- Allow a configurable portion of the tokens to be transfer right away
- Create a new token vesting contract with the user as the beneficiary
- Assign user as the owner (in the ERC-721 terms) of the new vesting contract
- Implement new logic to handle transfer of vesting contracts as NFTs

The main question that remains open is which vesting contract we would like to use. The two current options are: 
- [VestingVaults](https://github.com/tapmydata/tap-protocol/blob/main/contracts/VestingVault.sol): a new contract implementation that allow releases in batches of time and includes a battery of tests
- OpenZeppelin's [TokenVesting](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/release-v3.0.0/contracts/drafts/TokenVesting.sol): an old implementation that was already deprecated by the team. They found a few downsides on this implementation discussed here: https://github.com/OpenZeppelin/openzeppelin-contracts/issues/1214 

TODO:

- [ ] Decide which vesting implementation we will use

**Comment by lkngtn** [comment]
I don't have a strong preference between the two, generally not a fan of using deprecated code though so lean towards the VestingVaults implementation.
